File: Imaginengion/src/Platform/OpenGL/OpenGLTexture2D.cpp
#include "impch.h"
#ifdef IMAGINE_OPENGL
#include "Renderer/Texture.h"

#include <stb_image.h>

#include <glad/glad.h>

namespace IM {

	namespace {
		GLenum _InternalFormat, _DataFormat;
	}
	
	Texture2D::Texture2D(const std::string& path) {
		IMAGINE_PROFILE_FUNCTION();

		int width, height, channels;
		stbi_set_flip_vertically_on_load(1);
		stbi_uc* data = nullptr;
		{
			IMAGINE_PROFILE_SCOPE("stbi_load - OpenGLTexture2D::OpenGLTexture2D(const std::string&)");
			data = stbi_load(path.c_str(), &width, &height, &channels, 0);

		}
		IMAGINE_CORE_ASSERT(data, "Failed to load image in OpenGLTexture2D constructor!");
		_Width = width;
		_Height = height;

		GLenum internalFormat = 0, dataFormat = 0;

		if (channels == 4) {
			internalFormat = GL_RGBA8;
			dataFormat = GL_RGBA;
		}
		else if (channels == 3) {
			internalFormat = GL_RGB8;
			dataFormat = GL_RGB;
		}

		IMAGINE_CORE_ASSERT(internalFormat, "Texture format not supported in OpenGLTexture2D: {}");
		IMAGINE_CORE_ASSERT(dataFormat, "Texture format not supported in OpenGLTexture2D: {}");

		_InternalFormat = internalFormat;
		_DataFormat = dataFormat;

		glCreateTextures(GL_TEXTURE_2D, 1, &_TextureID);
		glTextureStorage2D(_TextureID, 1, internalFormat, _Width, _Height);

		glTextureParameteri(_TextureID, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
		glTextureParameteri(_TextureID, GL_TEXTURE_MAG_FILTER, GL_NEAREST);

		glTextureParameteri(_TextureID, GL_TEXTURE_WRAP_S, GL_REPEAT);
		glTextureParameteri(_TextureID, GL_TEXTURE_WRAP_T, GL_REPEAT);

		glTextureSubImage2D(_TextureID, 0, 0, 0, _Width, _Height, dataFormat, GL_UNSIGNED_BYTE, data);

		stbi_image_free(data);
	}

	Texture2D::Texture2D(uint32_t width, uint32_t height) 
		: _Width(width), _Height(height){

		IMAGINE_PROFILE_FUNCTION();

		_InternalFormat = GL_RGBA8;
		_DataFormat = GL_RGBA;

		glCreateTextures(GL_TEXTURE_2D, 1, &_TextureID);
		glTextureStorage2D(_TextureID, 1, _InternalFormat, _Width, _Height);

		glTextureParameteri(_TextureID, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
		glTextureParameteri(_TextureID, GL_TEXTURE_MAG_FILTER, GL_NEAREST);

		glTextureParameteri(_TextureID, GL_TEXTURE_WRAP_S, GL_REPEAT);
		glTextureParameteri(_TextureID, GL_TEXTURE_WRAP_T, GL_REPEAT);
	}
	
	Texture2D::~Texture2D() {
		IMAGINE_PROFILE_FUNCTION();

		glDeleteTextures(1, &_TextureID);
	}
	
	void Texture2D::SetData(void* data, uint32_t size) {
		IMAGINE_PROFILE_FUNCTION();

		glTextureSubImage2D(_TextureID, 0, 0, 0, _Width, _Height, _DataFormat, GL_UNSIGNED_BYTE, data);
	}
	
	void Texture2D::Bind(uint32_t slot) const {
		IMAGINE_PROFILE_FUNCTION();

		glBindTextureUnit(slot, _TextureID);
	}
	
	void Texture2D::Unbind(uint32_t slot) const {
		glBindTexture(GL_TEXTURE_2D, 0);
	}
}
#endif

File: src/Textures/OpenGLTexture2D.zig
const std = @import("std");
const glad = @import("../Core/CImports.zig").glad;
const stb = @import("../Core/CImports.zig").stb;
const OpenGLTexture2D = @This();

_Width: c_int,
_Height: c_int,
_TextureID: c_uint,
_InternalFormat: c_uint,
_DataFormat: c_uint,

pub fn InitData(width: u32, height: u32, channels: u32, data: *anyopaque) OpenGLTexture2D {
    var internal_format: c_uint = 0;
    var data_format: c_uint = 0;
    if (channels == 4) {
        internal_format = glad.GL_RGBA8;
        data_format = glad.GL_RGBA;
    } else if (channels == 3) {
        internal_format = glad.GL_RGB8;
        data_format = glad.GL_RGB;
    } else {
        std.log.err("Textureformat not supported in OpenGLTexture2D when loading loading data", .{});
        @panic("");
    }

    var new_texture_id: c_uint = 0;
    glad.glCreateTextures(glad.GL_TEXTURE_2D, 1, &new_texture_id);
    glad.glTextureStorage2D(new_texture_id, 1, internal_format, @intCast(width), @intCast(height));

    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_MIN_FILTER, glad.GL_LINEAR);
    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_MAG_FILTER, glad.GL_NEAREST);

    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_WRAP_S, glad.GL_REPEAT);
    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_WRAP_T, glad.GL_REPEAT);
    glad.glTextureSubImage2D(new_texture_id, 0, 0, 0, @intCast(width), @intCast(height), data_format, glad.GL_UNSIGNED_BYTE, data);

    return OpenGLTexture2D{
        ._Width = @intCast(width),
        ._Height = @intCast(height),
        ._TextureID = new_texture_id,
        ._InternalFormat = internal_format,
        ._DataFormat = data_format,
    };
}

pub fn InitPath(path: []const u8) !OpenGLTexture2D {
    var width: c_int = 0;
    var height: c_int = 0;
    var channels: c_int = 0;
    var data: ?*stb.stbi_uc = null;
    stb.stbi_set_flip_vertically_on_load(1);

    var file = try std.fs.openFileAbsolute(path, .{});
    defer file.close();
    const fstats = try file.stat();

    const contents = try file.readToEndAlloc(std.heap.page_allocator, @intCast(fstats.size));
    defer std.heap.page_allocator.free(contents);

    data = stb.stbi_load_from_memory(contents.ptr, @intCast(contents.len), &width, &height, &channels, 0);
    defer stb.stbi_image_free(data);
    std.debug.assert(data != null);

    var internal_format: c_uint = 0;
    var data_format: c_uint = 0;
    if (channels == 4) {
        internal_format = glad.GL_RGBA8;
        data_format = glad.GL_RGBA;
    } else if (channels == 3) {
        internal_format = glad.GL_RGB8;
        data_format = glad.GL_RGB;
    } else {
        std.log.err("Textureformat not supported in OpenGLTexture2D when loading file: {s}", .{path});
        @panic("");
    }

    var new_texture_id: c_uint = 0;
    glad.glCreateTextures(glad.GL_TEXTURE_2D, 1, &new_texture_id);
    glad.glTextureStorage2D(new_texture_id, 1, internal_format, width, height);

    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_MIN_FILTER, glad.GL_LINEAR);
    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_MAG_FILTER, glad.GL_NEAREST);

    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_WRAP_S, glad.GL_REPEAT);
    glad.glTextureParameteri(new_texture_id, glad.GL_TEXTURE_WRAP_T, glad.GL_REPEAT);

    glad.glTextureSubImage2D(new_texture_id, 0, 0, 0, width, height, data_format, glad.GL_UNSIGNED_BYTE, data);

    return OpenGLTexture2D{
        ._Width = width,
        ._Height = height,
        ._TextureID = new_texture_id,
        ._InternalFormat = internal_format,
        ._DataFormat = data_format,
    };
}

pub fn Deinit(self: OpenGLTexture2D) void {
    glad.glDeleteTextures(1, &self._TextureID);
}
pub fn GetWidth(self: OpenGLTexture2D) u32 {
    return self._Width;
}
pub fn GetHeight(self: OpenGLTexture2D) u32 {
    return self._Height;
}
pub fn GetID(self: OpenGLTexture2D) u32 {
    return self._TextureID;
}
pub fn UpdateData(self: *OpenGLTexture2D, width: u32, height: u32, data: *anyopaque, size: usize) void {
    _ = size;
    self._Width = width;
    self._Height = height;
    glad.glTextureSubImage2D(self._TextureID, 0, 0, 0, self._Width, self._Height, self._DataFormat, glad.GL_UNSIGNED_BYTE, data);
}
pub fn UpdateDataPath(self: *OpenGLTexture2D, path: []const u8) !void {
    var width: c_int = 0;
    var height: c_int = 0;
    var channels: c_int = 0;
    var data: ?*stb.stbi_uc = null;
    stb.stbi_set_flip_vertically_on_load(1);

    var file = try std.fs.openFileAbsolute(path, .{});
    defer file.close();
    const fstats = try file.stat();

    const contents = try file.readToEndAlloc(std.heap.page_allocator, @intCast(fstats.size));
    defer std.heap.page_allocator.free(contents);

    data = stb.stbi_load_from_memory(contents.ptr, @intCast(contents.len), &width, &height, &channels, 0);
    defer stb.stbi_image_free(data);
    std.debug.assert(data != null);

    self._Width = width;
    self._Height = height;
    glad.glTextureSubImage2D(self._TextureID, 0, 0, 0, self._Width, self._Height, self._DataFormat, glad.GL_UNSIGNED_BYTE, data);
}
pub fn Bind(self: OpenGLTexture2D, slot: u32) void {
    glad.glBindTextureUnit(slot, self._TextureID);
}
pub fn Unbind(self: OpenGLTexture2D, slot: u32) void {
    _ = self;
    _ = slot;
    glad.glBindTexture(glad.GL_TEXTURE_2D, 0);
}
